internal/backend/sessions/sessionworkflows/syscalls.go

```diff
@@ -211,6 +228,99 @@ func (w *sessionWorkflow) nextEvent(ctx context.Context, args []sdktypes.Value,
 	}
 }
 
+func (w *sessionWorkflow) join(ctx context.Context, args []sdktypes.Value, kwargs map[string]sdktypes.Value) (sdktypes.Value, error) {
+	var timeout time.Duration
+	if err := sdkmodule.UnpackArgs(nil, kwargs, "timeout=?", &timeout); err != nil {
+		return sdktypes.InvalidValue, err
+	}
+
+	if len(args) == 0 {
+		return sdktypes.InvalidValue, errors.New("expecting at least one signal")
+	}
+
+	if timeout < 0 {
+		return sdktypes.InvalidValue, errors.New("timeout must be a non-negative value")
+	}
+
+	signals, err := kittehs.ListToMapError(args, func(v sdktypes.Value) (uuid.UUID, bool, error) {
+		s := v.GetString().Value()
+
+		if sid, err := sdktypes.ParseSessionID(s); err == nil {
+			return sid.UUIDValue(), true, nil
+		}
+
+		id, err := uuid.Parse(s)
+		if err != nil {
+			return uuid.Nil, false, err
+		}
+
+		return id, true, nil
+	})
+	if err != nil {
+		return sdktypes.InvalidValue, err
+	}
+
+	received := make(map[string]sdktypes.Value, len(signals))
+
+	wctx := sessioncontext.GetWorkflowContext(ctx)
+
+	var timeoutFuture workflow.Future
+	if timeout != 0 {
+		timeoutFuture = workflow.NewTimer(wctx, timeout)
+	}
+
+L:
+	for len(signals) > 0 {
+		// check if there is an event already in one of the signals
+		for signalID := range signals {
+			event, err := w.getNextEvent(ctx, signalID)
+			if err != nil {
+				return sdktypes.InvalidValue, err
+			}
+
+			if event == nil {
+				continue
+			}
+
+			if received[signalID.String()], err = sdktypes.WrapValue(event); err != nil {
+				return sdktypes.InvalidValue, err
+			}
+
+			delete(signals, signalID)
+			continue L
+		}
+
+		for {
+			// no event, wait for first signal
+			signalID, err := w.waitOnFirstSignal(wctx, slices.Collect(maps.Keys(signals)), timeoutFuture)
+			if err != nil {
+				return sdktypes.InvalidValue, err
+			}
+
+			if signalID == uuid.Nil {
+				// timeout.
+				break L
+			}
+
+			// get next event on this signal
+			event, err := w.getNextEvent(ctx, signalID)
+			if err != nil {
+				return sdktypes.InvalidValue, err
+			}
+
+			if received[signalID.String()], err = sdktypes.WrapValue(event); err != nil {
```

Comment:
This is confusing, I prefer:
```suggestion

			wrapped, err := sdktypes.WrapValue(event)
			if err != nil { .... }
			received[signalID.String()] = wrapped
```